library/ajax/document_helpers.php

```diff
@@ -23,13 +23,22 @@
 
 require_once(dirname(__FILE__) . "/../../interface/globals.php");
 
-$term = isset($_GET["term"]) ? filter_input(INPUT_GET, 'term') : '';
+$req = array(
+    'term' => (isset($_GET["term"]) ? filter_input(INPUT_GET, 'term') : ''),
+    'sql_limit' => (isset($_GET["limit"]) ? filter_input(INPUT_GET, 'limit') : 20),
+);
 
-function get_patients_list($term)
+function get_patients_list($req)
 {
-    $term = "%" . $term . "%";
+    $term = "%" . $req['term'] . "%";
     $clear = "- " . xl("Reset to no patient") . " -";
-    $response = sqlStatement("SELECT Concat(patient_data.fname, ' ',patient_data.lname) as label, patient_data.pid as value FROM patient_data HAVING label LIKE ? ORDER BY patient_data.lname ASC", array($term));
+    $response = sqlStatement("
+        SELECT CONCAT(fname, ' ',lname,IF(IFNULL(deceased_date,0)=0,'','*')) as label, pid as value
+            FROM patient_data 
+            HAVING label LIKE ?
+            ORDER BY IF(IFNULL(deceased_date,0)=0, 0, 1) ASC, IFNULL(deceased_date,0) DESC, lname ASC, fname ASC
+            LIMIT ?", 
```

Comment:
You can't bind a limit. instead surround it with the specific limit escaping function:
https://github.com/openemr/openemr/blob/master/library/formdata.inc.php#L49

Reply:
It may be safer to provide a central sqlSelectLimit function that implements standard [selectLimit](http://adodb.org/dokuwiki/doku.php?id=v5:reference:connection:selectlimit) method.

Reply:
This is really odd, because the above bind should break. Is this working on testing?

Reply:
Generally need to use:
```
LIMIT " . escape_limit($req['sql_limit']),
```

Reply:
Also much better to use the function in the actual query itself to prevent a developer from mistakenly using escaping code in coding algorithms (this is why avoid ever doing sql escaping processing at top of scripts)

Reply:
Or if the binding does actually work, then do the escape_limit function in the line below:
```
array($term, escape_limit($req['sql_limit']))); 
```

Reply:
1. $binds should account for type - if the default is set as '20', it should generate sql error.  

2. The documentation from vendor does not warn against binding LIMIT for the execute function.  However using selectLimit eliminates errors due to database specific LIMIT clause variations.  

3. It is better to clean up / validate at the source as long as that variable has specific purpose.  As more functions are added to this library, developers should not be required to perform validations before every use - specially for integers.

Reply:
If you have a valid session, use following quick tests for binding:
1. Enter url - <web root>/library/ajax/document_helpers.php?limit=a
Response: [{"label":"- Reset to no patient -","value":"00"}]
2. Enter url - <web root>/library/ajax/document_helpers.php
Response: json object with 20 or less label objects
3. Enter url - <web root>/library/ajax/document_helpers.php?limit=100
Response: json object with 100 or less label objects
4. Enter url - <web root>/library/ajax/document_helpers.php?limit=-20
Response: MySQL server version for the right syntax to use near '-20' at line 5 (Bug in escape_limit)

Something to be noted here - this type of testing for background calls is possible because current code uses GETs.  

This PR was intended as a quick fix to avoid server load sending names of every patient and causing a noticeable lags with every keystroke,  Although I think last commit with escape_limit is the appropriate way for its intended purpose, if it helps I could replace it with sanitizer.  Unfortunately we won't be able to spend time on implementing any error handling.

Thx.

Reply:
I'm getting ready to do PR for documents use too limit to 3 char before search but a return limit is much needed i'd say 20 or so is more than enough.

Reply:
@mdsupport , The only goal of the sql escaping functions are to prevent sql injection. Since we can bind limit, makes sense to use this for primary escaping method (and will use escape_limit in cases when can not bind the limit such as really complicated sql query builds). So, we should switch this to the sanitize function at the top (I am happy to do this to avoid delaying this PR especially since @sjpadgett is working on related code also).

Reply:
Also to quickly add why I want to avoid a "sql escaping" functions at the top of the script; in the past, this makes it much more complicated to ensure scripts are adequately sql escaping (but when it is done as far downstream as possible, then it is clear that variables are adequately sql escaped); in this case, a moot point, since the limit is going through the binding mechanism. btw, use same logic for html escaping.